src/box/system_allocator.h

```diff
@@ -0,0 +1,181 @@
+ *    copyright notice, this list of conditions and the following
+ *    disclaimer in the documentation and/or other materials
+ *    provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY <COPYRIGHT HOLDER> ``AS IS'' AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
+ * TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL
+ * <COPYRIGHT HOLDER> OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT,
+ * INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR
+ * BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+ * LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF
+ * THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
+ * SUCH DAMAGE.
+ */
+#include <stdlib.h>
+#include <trivia/util.h>
+
+#if defined(__cplusplus)
+extern "C" {
+#endif /* defined(__cplusplus) */
+
+#if (defined (__APPLE__) || defined (__MACH__))
+#include <malloc/malloc.h>
+static inline size_t
+portable_malloc_usable_size(void *p)
+{
+	return malloc_size(p);
+}
+#elif (defined(__FreeBSD__) || defined(__NetBSD__) || \
+       defined(__OpenBSD__) || defined(__bsdi__))
+#include <malloc_np.h>
+static inline size_t
+portable_malloc_usable_size(void *p)
+{
+	return malloc_usable_size(p);
+}
+#elif defined(__linux__)
+#include <malloc.h>
+static inline size_t
+portable_malloc_usable_size(void *p)
+{
+	return malloc_usable_size(p);
+}
+#else
+#error "Undefined system type"
+#endif
+
+
+struct system_alloc {
+	/**
+	 * Bytes allocated by system allocator
+	 */
+	uint64_t used_bytes;
+	/**
+	 * Allocator quota
+	 */
+	struct quota *quota;
+};
+
+struct system_stats {
+	size_t used;
+	size_t total;
+};
+
+enum system_opt {
+	SYSTEM_DELAYED_FREE_MODE
+};
+
+typedef int (*system_stats_cb)(const struct system_stats *stats,
+				void *cb_ctx);
+
+static inline void
+system_alloc_setopt(struct system_alloc *alloc, enum system_opt opt, bool val)
+{
+	(void)alloc;
+	(void)opt;
+	(void)val;
+}
+
+static inline void
+system_stats(struct system_alloc *alloc, struct system_stats *totals,
+	mempool_stats_cb cb, void *cb_ctx)
+{
+	struct mempool_stats stats;
+	memset(&stats, 0, sizeof(stats));
+	totals->used =  pm_atomic_load_explicit(&alloc->used_bytes,
+		pm_memory_order_relaxed);
+	totals->total = quota_total(alloc->quota);
+	stats.totals.used = totals->used;
+	stats.totals.total = totals->total;
+	cb(&stats, cb_ctx);
+}
+
+static inline void
+system_alloc_create(struct system_alloc *alloc, struct quota *quota,
+	uint32_t objsize_min, float alloc_factor)
+{
+	alloc->used_bytes = 0;
+	alloc->quota = quota;
+	(void)objsize_min;
+	(void)alloc_factor;
+}
+
+static inline void
+system_alloc_destroy(MAYBE_UNUSED struct system_alloc *alloc)
+{
+
```

Comment:
So if sysfree() will happen later, it will crash, because alloc may be already deleted? I suggest you to add a panic here in case not everything is freed, maybe.

Reply:
As i see in small_alloc_destroy we not check that all memory was free, before we destroy allocator. I add some debug info in memtx_engine_shutdown function (i call small_stats and print result), where we destroy allocator and see that not all memory was freed (in master branch also). So i add init flag which set in system_allocator_init and reset in system_allocator_destroy, also i add asserts in all allocator functions to check that flag is set (i check that allocator not used before init and after destroy). But the full release of memory remains on the conscience of the allocator user. How do you think its ok?